### PR TITLE
Added some advanced YAML examples

### PR DESCRIPTION
**Pull request by Graeme-Miller**
Added some advanced YAML examples and made a CollectionFunctionals constructor public

### CODE REVIEW

File: docs/guide/yaml/advanced-example.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
+---
+title: YAML BLueprint Advanced Example
+layout: website-normal
+---
+
+By this point you should be familiar with the fundamental concepts behind both Apache Brooklyn and YAML blueprints. This section of the documentation is intended to show a complete, advanced example of a YAML blueprint.
+
+The intention is that this example is used to learn the more in-depth concepts, but also to serve as a reference point when writing your own blueprints. This page will first exaplin what the example is and how to run it, then it will spotlight interesting features.
```
Comment (sjcorbett) [inline docs/guide/yaml/advanced-example.md]: `exaplin`

File: docs/guide/yaml/advanced-example.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+There are four blueprints that make up this application. Each of them are used to add one or more catalog items to Brooklyn. You can find them below:
+
+* [Elasticsearch](example_yaml/brooklyn-elasticsearch-catalog.bom)
+* [Logstash](example_yaml/brooklyn-logstash-catalog.bom)
+* [Kibana](example_yaml/brooklyn-kibana-catalog.bom)
+* [ELK](example_yaml/brooklyn-elk-catalog.bom)
+
+#### Running the example
+First, add all four blueprints to the Brooklyn Catalog. This can be done by clicking the 'Catalog' tab, clinking the '+' symbol and pasting the YAML. Once this is done , click the 'Application' then the '+' button to bring up the add application wizard. A new Catalog item will be available called 'ELK Stack'. Using the add application wizard, you should be able to deploy an ELK stack to a location of your choosing.
```
Comment (sjcorbett) [inline docs/guide/yaml/advanced-example.md]: `done , click`

Comment (sjcorbett) [inline docs/guide/yaml/advanced-example.md]: `the 'Application'` what?

File: docs/guide/yaml/advanced-example.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+~~~yaml
+brooklyn.enrichers:
+  - type: org.apache.brooklyn.enricher.stock.Transformer
+    brooklyn.config:
+      enricher.sourceSensor: $brooklyn:sensor("host.address")
+      enricher.targetSensor: $brooklyn:sensor("url.tcp")
+      enricher.targetValue: $brooklyn:formatString("%s:%s", $brooklyn:attributeWhenReady("host.address"), $brooklyn:config("elasticsearch.tcp.port"))  
+~~~
+
+In this example, we take the host.address and append the TCP port, outputting the result as url.tcp. After this has been done, we now need to collet all the URLs into a list in the Cluster entity, as follows:
```
Comment (sjcorbett) [inline docs/guide/yaml/advanced-example.md]: `collet`

File: docs/guide/yaml/advanced-example.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+~~~yaml
+brooklyn.enrichers:
+  - type: org.apache.brooklyn.enricher.stock.Aggregator
+    brooklyn.config:
+      enricher.sourceSensor: $brooklyn:sensor("url.tcp")
+      enricher.targetSensor: $brooklyn:sensor("urls.tcp.list")
+      enricher.aggregating.fromMembers: true
+
+~~~
+In the proceeding example, we aggregate all of the TCP urls generated in the early example. These are then stored in a sensor called urls.tcp.list. This list is then joined together into one long string:
```
Comment (sjcorbett) [inline docs/guide/yaml/advanced-example.md]: s/proceeding/preceding/

File: docs/guide/yaml/example_yaml/brooklyn-elasticsearch-catalog.bom [hunk 5]
```diff
@@ -0,0 +1,126 @@
+brooklyn.catalog:
+  id: elasticsearch
+  version: 1.0
+  iconUrl: https://avatars0.githubusercontent.com/u/6764390?v=3&s=400
+  name: Elasticsearch
+  license: Apache-2.0
+  maintainer_name: Graeme Miller
+  maintainer_email: graeme.miller@cloudsoftcorp.com
```
Comment (sjcorbett) [inline docs/guide/yaml/example_yaml/brooklyn-elasticsearch-catalog.bom]: Would prefer to exclude maintainer name/email for the same reason as `@author` tags are discouraged.

File: utils/common/src/main/java/org/apache/brooklyn/util/collections/CollectionFunctionals.java [hunk 6]
```diff
@@ -107,12 +107,12 @@ public Integer apply(Iterable<?> input) {
...
 
     public static final class FirstElementFunction<T> implements Function<Iterable<? extends T>, T> {
-        private FirstElementFunction() {
+        public FirstElementFunction() {
         }
 
         @Override
         public T apply(Iterable<? extends T> input) {
-            if (input==null) return null;
+            if (input==null || Iterables.isEmpty(input)) return null;
```
Comment (ahgittin) [inline utils/common/src/main/java/org/apache/brooklyn/util/collections/CollectionFunctionals.java]: good addition!

File: utils/common/src/main/java/org/apache/brooklyn/util/collections/CollectionFunctionals.java [hunk 7]
```diff
@@ -107,12 +107,12 @@ public Integer apply(Iterable<?> input) {
     }
 
     public static final class FirstElementFunction<T> implements Function<Iterable<? extends T>, T> {
-        private FirstElementFunction() {
+        public FirstElementFunction() {
```
Comment (ahgittin) [inline utils/common/src/main/java/org/apache/brooklyn/util/collections/CollectionFunctionals.java]: is this needed?  we have a static method `CollectionFunctionals.firstElement()` to return the instance, and the class is `final` so you can't subclass so i'm not sure what benefit there is from making this public.  (a javadoc on the constructor pointing at the method could be useful though!)

no strong feelings though

**Comment by ahgittin** [comment]
editorial comments from @sjcorbett and minor comment from me, otherwise good to merge

nice to have these, at least in the short term

longer term i'd like them all in a hosted catalog, maybe fewer here -- and those which are here we need to make part of a QA regime!!!

File: utils/common/src/main/java/org/apache/brooklyn/util/collections/CollectionFunctionals.java (hunk 7 above)
Comment (sjcorbett) [inline utils/common/src/main/java/org/apache/brooklyn/util/collections/CollectionFunctionals.java]: I suspect it's necessary to use it from YAML:
```
     enricher.transformation:
     $brooklyn:object:
       type: "org.apache.brooklyn.util.collections.CollectionFunctionals$FirstElementFunction"
```

Comment (Graeme-Miller) [inline utils/common/src/main/java/org/apache/brooklyn/util/collections/CollectionFunctionals.java]: yup, it is needed to use it from YAML.

**Comment by sjcorbett** [comment]
Jenkins failure caused by 1075. Will merge this.

**Pull request merged**